File: ads/graph/crud.py
import inspect
from contextlib import contextmanager
from typing import List, Optional

from sqlalchemy.orm import Session

from .models import Node, Edge
from ..storage import database
from ..storage.database import get_db


class GraphCRUD:
    """图谱 CRUD 操作，支持可注入的 SQLAlchemy 会话。"""

    @staticmethod
    @contextmanager
    def _session_scope(session: Optional[Session] = None):
        if session is not None:
            yield session, False
            return

        session_factory = getattr(database, "_SessionLocal", None)
        if session_factory is None:
            attr_factory = getattr(database, "SessionLocal", None)
            if attr_factory is not None and not inspect.isfunction(attr_factory):
                session_factory = attr_factory

        if session_factory is not None:
            db = session_factory()
            try:
                yield db, True
                db.commit()
            except Exception:
                db.rollback()
                raise
            finally:
                db.close()
            return

        with get_db() as db:
            yield db, True

    @staticmethod
    def create_node(
        id: str,
        type: str,
        label: str,
        content: Optional[str] = "",
        metadata: Optional[dict] = None,
        position: Optional[dict] = None,
        workspace_id: Optional[int] = None,
        is_draft: Optional[bool] = False,
        session: Optional[Session] = None,
    ) -> Node:
        """创建节点，支持外部会话。"""

        metadata = metadata or {}
        position = position or {"x": 0, "y": 0}
        has_content = bool(content and str(content).strip())
        stored_content = content
        draft_value = None

        if is_draft is None:
            is_draft_value = not has_content
        else:
            is_draft_value = is_draft

        if is_draft_value and has_content:
            draft_value = stored_content

        with GraphCRUD._session_scope(session) as (db, managed):
            node = Node(
                id=id,
                type=type,
                label=label,
                content=stored_content,
                draft_content=draft_value,
                is_draft=is_draft_value,
                current_version=1 if has_content and not is_draft_value else 0,
                node_metadata=metadata,
                position=position,
            )
            db.add(node)
            db.flush()
            if managed:
                db.refresh(node)
            return node

    @staticmethod
    def get_all_nodes(
        workspace_id: Optional[int] = None,
        filter_by_workspace: bool = False,
        session: Optional[Session] = None,
    ) -> List[Node]:
        """获取所有节点。"""

        with GraphCRUD._session_scope(session) as (db, _):
            query = db.query(Node)

            if filter_by_workspace:
                if workspace_id is not None:
                    query = query.filter(Node.workspace_id == workspace_id)
                else:
                    query = query.filter(Node.workspace_id == None)  # noqa: E711

            nodes = query.all()
            for node in nodes:
                _ = (
                    node.id,
                    node.type,
                    node.label,
                    node.content,
                    node.node_metadata,
                    node.position,
                    node.created_at,
                    node.updated_at,
                )
            return nodes

    @staticmethod
    def get_node_by_id(node_id: str, session: Optional[Session] = None) -> Optional[Node]:
        """根据 ID 获取节点。"""

        with GraphCRUD._session_scope(session) as (db, _):
            node = db.query(Node).filter(Node.id == node_id).first()
            if node:
                _ = (
                    node.id,
                    node.type,
                    node.label,
                    node.content,
                    node.node_metadata,
                    node.position,
                    node.created_at,
                    node.updated_at,
                    node.current_version,
                    node.draft_content,
                    node.draft_source_type,
                    node.draft_conversation_id,
                    node.draft_message_id,
                    node.draft_based_on_version,
                    node.draft_ai_original_content,
                    node.is_draft,
                    node.draft_updated_at,
                )
            return node

    @staticmethod
    def update_node(
        node_id: str,
        updates: dict,
        session: Optional[Session] = None,
    ) -> Optional[Node]:
        """更新节点。"""

        with GraphCRUD._session_scope(session) as (db, managed):
            node = db.query(Node).filter(Node.id == node_id).first()
            if not node:
                return None

            updates = dict(updates or {})
            if "metadata" in updates:
                updates["node_metadata"] = updates.pop("metadata")

            for key, value in updates.items():
                if hasattr(node, key):
                    setattr(node, key, value)

            db.flush()
            if managed:
                db.refresh(node)
            return node

    @staticmethod
    def delete_node(node_id: str, session: Optional[Session] = None) -> bool:
        """删除节点。"""

        with GraphCRUD._session_scope(session) as (db, _):
            node = db.query(Node).filter(Node.id == node_id).first()
            if not node:
                return False
            db.delete(node)
            db.flush()
            return True

    @staticmethod
    def create_edge(
        id: str,
        source: str,
        target: str,
        label: str = "",
        edge_type: str = "next",
        animated: bool = False,
        source_handle: str = "right",
        target_handle: str = "left",
        workspace_id: Optional[int] = None,
        session: Optional[Session] = None,
    ) -> Edge:
        """创建边。"""

        with GraphCRUD._session_scope(session) as (db, managed):
            edge = Edge(
                id=id,
                source=source,
                target=target,
                source_handle=source_handle,
                target_handle=target_handle,
                label=label,
                edge_type=edge_type,
                animated=animated,
            )
            db.add(edge)
            db.flush()
            if managed:
                db.refresh(edge)
            return edge

    @staticmethod
    def get_all_edges(
        workspace_id: Optional[int] = None,
        filter_by_workspace: bool = False,
        session: Optional[Session] = None,
    ) -> List[Edge]:
        """获取所有边。"""

        with GraphCRUD._session_scope(session) as (db, _):
            query = db.query(Edge)

            if filter_by_workspace:
                if workspace_id is not None:
                    query = query.filter(Edge.workspace_id == workspace_id)
                else:
                    query = query.filter(Edge.workspace_id == None)  # noqa: E711

            edges = query.all()
            for edge in edges:
                _ = (
                    edge.id,
                    edge.source,
                    edge.target,
                    edge.source_handle,
                    edge.target_handle,
                    edge.label,
                    edge.edge_type,
                    edge.animated,
                    edge.created_at,
                    edge.updated_at,
                )
            return edges

    @staticmethod
    def delete_edge(edge_id: str, session: Optional[Session] = None) -> bool:
        """删除边。"""

        with GraphCRUD._session_scope(session) as (db, _):
            edge = db.query(Edge).filter(Edge.id == edge_id).first()
            if not edge:
                return False
            db.delete(edge)
            db.flush()
            return True

    @staticmethod
    def get_edges_from_node(
        node_id: str,
        session: Optional[Session] = None,
    ) -> List[Edge]:
        """获取从指定节点出发的所有边。"""

        with GraphCRUD._session_scope(session) as (db, _):
            edges = db.query(Edge).filter(Edge.source == node_id).all()
            for edge in edges:
                _ = (
                    edge.id,
                    edge.source,
                    edge.target,
                    edge.source_handle,
                    edge.target_handle,
                    edge.label,
                    edge.edge_type,
                    edge.animated,
                    edge.created_at,
                    edge.updated_at,
                )
            return edges

    @staticmethod
    def get_next_node(node_id: str, session: Optional[Session] = None) -> Optional[Node]:
        """获取下一个节点（根据 next 类型的边）。"""

        with GraphCRUD._session_scope(session) as (db, _):
            edge = (
                db.query(Edge)
                .filter(Edge.source == node_id, Edge.edge_type == "next")
                .first()
            )
            if not edge:
                return None
            return GraphCRUD.get_node_by_id(edge.target, session=session)

    @staticmethod
    def get_parent_nodes(
        node_id: str,
        recursive: bool = True,
        session: Optional[Session] = None,
    ) -> List[Node]:
        """递归查找所有父节点。"""

        with GraphCRUD._session_scope(session) as (db, _):
            parents: List[Node] = []
            current_id = node_id
            seen_ids = set()

            while True:
                edge = (
                    db.query(Edge)
                    .filter(Edge.target == current_id, Edge.source != current_id)
                    .first()
                )
                if not edge:
                    break

                parent = db.query(Node).filter(Node.id == edge.source).first()
                if not parent or parent.id in seen_ids:
                    break

                _ = (
                    parent.id,
                    parent.type,
                    parent.label,
                    parent.content,
                    parent.node_metadata,
                    parent.position,
                    parent.created_at,
                    parent.updated_at,
                )

                parents.append(parent)
                seen_ids.add(parent.id)

                if not recursive:
                    break

                current_id = edge.source

            return parents

    @staticmethod
    def get_node_context(
        node_id: str,
        session: Optional[Session] = None,
    ) -> Optional[dict]:
        """获取节点及其上下文信息（包含父节点）。"""

        node = GraphCRUD.get_node_by_id(node_id, session=session)
        if not node:
            return None

        parents = GraphCRUD.get_parent_nodes(node_id, recursive=True, session=session)

        return {
            "node": node,
            "parents": parents,
        }

File: server/graph/crud.ts
import { getDatabase } from "../storage/database.js";
import { parseOptionalSqliteInt, parseSqliteBoolean, parseSqliteJsonObject } from "../utils/sqlite.js";
import type { GraphNode, GraphEdge } from "./types.js";

interface CreateNodeInput {
  id: string;
  type: string;
  label: string;
  content?: string | null;
  metadata?: Record<string, unknown>;
  position?: Record<string, unknown>;
  workspaceId?: number | null;
  isDraft?: boolean | null;
}

type NodeUpdateInput = Record<string, unknown>;

type SqlDateValue = string | number | Date | null | undefined;

export interface NodeRow {
  id: string;
  type: string;
  label: string;
  content?: string | null;
  metadata?: string | Record<string, unknown> | null;
  position?: string | Record<string, unknown> | null;
  current_version?: number | null;
  draft_content?: string | null;
  draft_source_type?: string | null;
  draft_conversation_id?: string | null;
  draft_message_id?: number | string | null;
  draft_based_on_version?: number | null;
  draft_ai_original_content?: string | null;
  draft_updated_at?: SqlDateValue;
  is_draft?: number | boolean | null;
  created_at?: SqlDateValue;
  updated_at?: SqlDateValue;
  workspace_id?: number | null;
}

interface EdgeRow {
  id: string;
  source: string;
  target: string;
  source_handle?: string | null;
  target_handle?: string | null;
  label?: string | null;
  edge_type: string;
  animated?: number | boolean | null;
  created_at?: SqlDateValue;
  updated_at?: SqlDateValue;
}

function normalizeNodeRow(raw: unknown): NodeRow {
  if (!raw || typeof raw !== "object") {
    throw new Error("Invalid node row: expected object");
  }
  const row = raw as Record<string, unknown>;
  if (typeof row.id !== "string" || !row.id) {
    throw new Error("Invalid node row: missing id");
  }
  if (typeof row.type !== "string" || !row.type) {
    throw new Error(`Invalid node row ${row.id}: missing type`);
  }

  return {
    ...(raw as NodeRow),
    id: row.id,
    type: row.type,
    label: typeof row.label === "string" ? row.label : String(row.label ?? ""),
  };
}

function normalizeEdgeRow(raw: unknown): EdgeRow {
  if (!raw || typeof raw !== "object") {
    throw new Error("Invalid edge row: expected object");
  }
  const row = raw as Record<string, unknown>;
  if (typeof row.id !== "string" || !row.id) {
    throw new Error("Invalid edge row: missing id");
  }
  if (typeof row.source !== "string" || !row.source) {
    throw new Error(`Invalid edge row ${row.id}: missing source`);
  }
  if (typeof row.target !== "string" || !row.target) {
    throw new Error(`Invalid edge row ${row.id}: missing target`);
  }
  if (typeof row.edge_type !== "string" || !row.edge_type) {
    throw new Error(`Invalid edge row ${row.id}: missing edge_type`);
  }

  return raw as EdgeRow;
}

function normalizeDate(value: SqlDateValue): Date | null {
  if (value === null || value === undefined) {
    return null;
  }
  if (value instanceof Date) {
    return value;
  }
  return new Date(value);
}

function mapNode(row: NodeRow): GraphNode {
  const metadata = parseSqliteJsonObject(row.metadata, {});
  const position = parseSqliteJsonObject(row.position, { x: 0, y: 0 });
  const draftMessageId = parseOptionalSqliteInt(row.draft_message_id);

  return {
    id: row.id,
    type: row.type,
    label: row.label,
    content: row.content ?? null,
    metadata,
    position,
    currentVersion: row.current_version ?? 0,
    draftContent: row.draft_content ?? null,
    isDraft: parseSqliteBoolean(row.is_draft),
    createdAt: normalizeDate(row.created_at),
    updatedAt: normalizeDate(row.updated_at),
    draftSourceType: row.draft_source_type ?? null,
    draftConversationId: row.draft_conversation_id ?? null,
    draftMessageId,
    draftBasedOnVersion: row.draft_based_on_version ?? null,
    draftAiOriginalContent: row.draft_ai_original_content ?? null,
    draftUpdatedAt: normalizeDate(row.draft_updated_at),
  };
}

function mapEdge(row: EdgeRow): GraphEdge {
  const animated =
    typeof row.animated === "boolean" ? row.animated : Number(row.animated) === 1;

  return {
    id: row.id,
    source: row.source,
    target: row.target,
    sourceHandle: row.source_handle ?? null,
    targetHandle: row.target_handle ?? null,
    label: row.label ?? null,
    edgeType: row.edge_type,
    animated,
    createdAt: normalizeDate(row.created_at),
    updatedAt: normalizeDate(row.updated_at),
  };
}

export function getAllNodes(): GraphNode[] {
  const db = getDatabase();
  const rows = db.prepare("SELECT * FROM nodes ORDER BY created_at ASC").all() as unknown[];
  return rows.map((row) => mapNode(normalizeNodeRow(row)));
}

export function getNodeById(nodeId: string): GraphNode | null {
  const db = getDatabase();
  const row = db.prepare("SELECT * FROM nodes WHERE id = ?").get(nodeId) as unknown;
  if (!row) {
    return null;
  }
  return mapNode(normalizeNodeRow(row));
}

export function createNode(input: CreateNodeInput): GraphNode {
  const db = getDatabase();
  const now = new Date().toISOString();

  const hasContent = typeof input.content === "string" && input.content.trim().length > 0;
  const isDraftValue =
    input.isDraft === undefined || input.isDraft === null
      ? !hasContent
      : Boolean(input.isDraft);
  const draftContent = isDraftValue && hasContent ? input.content ?? "" : null;
  const currentVersion = hasContent && !isDraftValue ? 1 : 0;

  const insert = db.prepare(`
    INSERT INTO nodes (
      id,
      type,
      label,
      content,
      metadata,
      position,
      current_version,
      draft_content,
      draft_source_type,
      draft_conversation_id,
      draft_message_id,
      draft_based_on_version,
      draft_ai_original_content,
      is_draft,
      draft_updated_at,
      created_at,
      updated_at,
      workspace_id
    ) VALUES (
      @id,
      @type,
      @label,
      @content,
      @metadata,
      @position,
      @current_version,
      @draft_content,
      NULL,
      NULL,
      NULL,
      NULL,
      NULL,
      @is_draft,
      NULL,
      @created_at,
      @updated_at,
      @workspace_id
    )
  `);

  insert.run({
    id: input.id,
    type: input.type,
    label: input.label,
    content: input.content ?? null,
    metadata: JSON.stringify(input.metadata ?? {}),
    position: JSON.stringify(input.position ?? { x: 0, y: 0 }),
    current_version: currentVersion,
    draft_content: draftContent,
    is_draft: isDraftValue ? 1 : 0,
    created_at: now,
    updated_at: now,
    workspace_id: input.workspaceId ?? null,
  });

  return getNodeById(input.id)!;
}

export function updateNode(nodeId: string, updates: NodeUpdateInput): GraphNode | null {
  if (!updates || Object.keys(updates).length === 0) {
    return getNodeById(nodeId);
  }

  const db = getDatabase();
  const fields: string[] = [];
  const params: Record<string, unknown> = { id: nodeId };

  const setField = (column: string, value: unknown): void => {
    fields.push(`${column} = @${column}`);
    params[column] = value;
  };

  for (const [key, value] of Object.entries(updates)) {
    switch (key) {
      case "label":
        setField("label", value);
        break;
      case "content":
        setField("content", value);
        break;
      case "metadata":
        setField("metadata", JSON.stringify(value ?? {}));
        break;
      case "position":
        setField("position", JSON.stringify(value ?? { x: 0, y: 0 }));
        break;
      case "current_version":
        setField("current_version", value);
        break;
      case "draft_content":
        setField("draft_content", value);
        break;
      case "is_draft":
        setField("is_draft", value ? 1 : 0);
        break;
      case "draft_source_type":
      case "draft_conversation_id":
      case "draft_message_id":
      case "draft_based_on_version":
      case "draft_ai_original_content":
      case "draft_updated_at":
        setField(key, value);
        break;
      default:
        break;
    }
  }

  if (fields.length === 0) {
    return getNodeById(nodeId);
  }

  fields.push("updated_at = @updated_at");
  params.updated_at = new Date().toISOString();

  const sql = `UPDATE nodes SET ${fields.join(", ")} WHERE id = @id`;
  db.prepare(sql).run(params);

  return getNodeById(nodeId);
}

export function deleteNode(nodeId: string): boolean {
  const db = getDatabase();
  const result = db.prepare("DELETE FROM nodes WHERE id = ?").run(nodeId);
  return result.changes > 0;
}

export function getAllEdges(): GraphEdge[] {
  const db = getDatabase();
  const rows = db.prepare("SELECT * FROM edges ORDER BY created_at ASC").all() as unknown[];
  return rows.map((row) => mapEdge(normalizeEdgeRow(row)));
}

export function createEdge(input: {
  id: string;
  source: string;
  target: string;
  edgeType?: string;
  label?: string | null;
  sourceHandle?: string | null;
  targetHandle?: string | null;
  animated?: boolean;
}): GraphEdge {
  const db = getDatabase();
  const now = new Date().toISOString();

  db.prepare(`
    INSERT INTO edges (
      id,
      source,
      target,
      source_handle,
      target_handle,
      label,
      edge_type,
      animated,
      created_at,
      updated_at
    ) VALUES (
      @id,
      @source,
      @target,
      @source_handle,
      @target_handle,
      @label,
      @edge_type,
      @animated,
      @created_at,
      @updated_at
    )
  `).run({
    id: input.id,
    source: input.source,
    target: input.target,
    source_handle: input.sourceHandle ?? "right",
    target_handle: input.targetHandle ?? "left",
    label: input.label ?? "",
    edge_type: input.edgeType ?? "next",
    animated: input.animated ? 1 : 0,
    created_at: now,
    updated_at: now,
  });

  const row = db.prepare("SELECT * FROM edges WHERE id = ?").get(input.id) as unknown;
  return mapEdge(normalizeEdgeRow(row));
}

export function deleteEdge(edgeId: string): boolean {
  const db = getDatabase();
  const result = db.prepare("DELETE FROM edges WHERE id = ?").run(edgeId);
  return result.changes > 0;
}

export function getEdgesFromNode(nodeId: string): GraphEdge[] {
  const db = getDatabase();
  const rows = db.prepare("SELECT * FROM edges WHERE source = ?").all(nodeId) as unknown[];
  return rows.map((row) => mapEdge(normalizeEdgeRow(row)));
}

export function getNextNode(nodeId: string): GraphNode | null {
  const db = getDatabase();
  const edge = db
    .prepare("SELECT target FROM edges WHERE source = ? AND edge_type = 'next' LIMIT 1")
    .get(nodeId) as { target?: string } | undefined;
  if (!edge?.target) {
    return null;
  }
  return getNodeById(edge.target);
}

export function getParentNodes(nodeId: string, recursive = true): GraphNode[] {
  const db = getDatabase();

  // Optimization: Use recursive CTE to fetch ancestors in a single query
  // instead of N+1 loop.
  // We use a path string to detect cycles and stop.
  // We use LIMIT 1 to mimic the original greedy "single path" behavior.

  if (!recursive) {
    const edge = db
      .prepare("SELECT source FROM edges WHERE target = ? AND source != ? LIMIT 1")
      .get(nodeId, nodeId) as { source?: string } | undefined;

    if (!edge?.source) {
      return [];
    }

    // Reuse getNodeById logic (which does a query + map)
    // Or we could join here too, but for depth 1 it's fine.
    const node = getNodeById(edge.source);
    return node ? [node] : [];
  }

  const sql = `
    WITH RECURSIVE lineage AS (
      -- Anchor: find all parents of the start node
      SELECT source, target, 1 as depth, '/' || target || '/' || source || '/' as path
      FROM edges
      WHERE target = ? AND source != target

      UNION ALL

      -- Recursive: find parents of the previous parents
      SELECT e.source, e.target, l.depth + 1, l.path || e.source || '/'
      FROM edges e
      JOIN lineage l ON e.target = l.source
      WHERE e.source != e.target
        AND l.depth < 100
        AND l.path NOT LIKE '%/' || e.source || '/%'
    )
    SELECT n.*
    FROM lineage l
    JOIN nodes n ON n.id = l.source
    GROUP BY n.id
    ORDER BY min(l.depth) ASC;
  `;

  const rows = db.prepare(sql).all(nodeId) as unknown[];
  return rows.map((row) => mapNode(normalizeNodeRow(row)));
}

export function getNodeContext(nodeId: string): { node: GraphNode; parents: GraphNode[] } | null {
  const node = getNodeById(nodeId);
  if (!node) {
    return null;
  }
  const parents = getParentNodes(nodeId, true);
  return { node, parents };
}
